### PR TITLE
Dev/ft/ring 28492 keyscheme enhancement

### PR DESCRIPTION
**Pull request by maximelubin**

### CODE REVIEW

**Comment by maximelubin** [comment]
First commits (RING-28696) are part of pull request https://github.com/scality/hdclient/pull/12.
Only those after should be reviewed

**Comment by maximelubin** [comment]
@bert-e after_pull_request=12

**Review by jgagnaire — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

File: lib/config.js [hunk 1]
```diff
@@ -74,59 +75,102 @@ function validatePolicySection(policy) {
...
-            configIsValid: false,
-            configError: new InvalidConfigError(
-                'codingParts',
-                opts.codingParts,
-                'Expected integer larger than 0'),
-        };
+    for (let i = 0; i < codes.length; i++) {
+        const pattern = codes[i];
+
+        if (typeof(pattern.dataParts) !== 'number' ||
```
Comment (xroche-scality) [inline lib/config.js]: C'est moche le JS :(

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

File: lib/config.js [hunk 2]
```diff
@@ -225,6 +225,11 @@ function validate(opts) {
         };
     }
 
+    if (!opts.serviceId) {
+        // eslint-disable-next-line no-param-reassign
+        opts.serviceId = 1; // Default
```
Comment (xroche-scality) [inline lib/config.js:230]: = placement.PLACEMENT_POLICY_VERSION ?

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

**Review by xroche-scality — APPROVED** [review]

File: Design.md [hunk 3]
```diff
@@ -48,10 +48,10 @@ Proposed key generation scheme:
 
 The keys actually used to sotre fragments on the hyperdrives can be derived easily with only the generated key, even for splits.
 ```
-<stored_fragment_key> := <serviceId>-<ctime>-<hash>-<start_offset>-<fragid>
+<stored_fragment_key> := <serviceId>-<ctime>-<hash>-<end_offset>-<fragid>
 <serviceId>, <ctime> and <hash> are the ones defined above
 <fragid>:= index in main key fragment list
-<start_offset> := used for splits. All split chunks share the same prefix, storing the offset is used to easily have range queries and avoid storing them all in the main key.
+<end_offset> := used for splits. All split chunks share the same prefix, storing the offset is used to easily have range queries and avoid storing them all in the main key. End of chunk offset is used to be able to infer object real size from last chunk keys alone (only way to do it for erasure coded last chunk parts).
```
Comment (xroche-scality) [inline Design.md]: Can we be more explicit that the end offset is EXCLUSIVE ? (ie. offset of byte just after the ending) - or did I misunderstand it ?

**Review by xroche-scality — APPROVED** [review]

File: lib/config.js (hunk 2 above)
Comment (maximelubin) [reply lib/config.js:230]: Nope. Serviceid is totally unrelated. Can be used for namspacing if/when we have other applications oustide of Zenko using hyperdrives

File: Design.md (hunk 3 above)
Comment (maximelubin) [reply Design.md]: It is indeed exclusive. WIll add a comment

File: lib/config.js (hunk 1 above)
Comment (maximelubin) [reply lib/config.js]: Agreed :( (but perhaps I am not doing it properly either)

**Pull request merged**